### PR TITLE
Store position of invalid char in `InvalidCharError`

### PR DESCRIPTION
**Pull request by leo-nash**
Fix: #98

### CODE REVIEW

File: src/iter.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@ pub type HexSliceToBytesIter<'a> = HexToBytesIter<HexDigitsIter<'a>>;
 
 /// Iterator yielding bytes decoded from an iterator of pairs of hex digits.
 pub struct HexToBytesIter<T: Iterator<Item = [u8; 2]>> {
-    iter: T,
+    iter: Enumerate<T>,
```
Comment (Kixunil) [inline src/iter.rs:18]: I think it'd be better to store the original length of the string and then compute position as `original_len - iter.len() * 2`. The reason is we probably want a fast success path and having to increase a number every loop iteration in addition to increasing the pointer inside iterator is waste of performance.

This needs `ExactSizeIterator` bound but that should be fine.

**Review by Kixunil — COMMENTED** [review]
Thanks for the PR! Even though I think it could be done differently, we might want to compare performance to test if my claim is correct.

File: src/iter.rs [hunk 2]
```diff
@@ -62,30 +70,50 @@ impl<T: Iterator<Item = [u8; 2]>> Iterator for HexToBytesIter<T> {
     #[inline]
     fn nth(&mut self, n: usize) -> Option<Self::Item> {
         let [hi, lo] = self.iter.nth(n)?;
-        Some(hex_chars_to_byte(hi, lo))
+        Some(hex_chars_to_byte(hi, lo).map_err(|(c, h)| InvalidCharError {
```
Comment (tcharding) [inline src/iter.rs:73]: nit: Perhaps use `(c, is_high)`

File: src/iter.rs [hunk 3]
```diff
@@ -153,9 +181,9 @@ impl<'a> ExactSizeIterator for HexDigitsIter<'a> {}
 impl<'a> core::iter::FusedIterator for HexDigitsIter<'a> {}
 
 /// `hi` and `lo` are bytes representing hex characters.
-fn hex_chars_to_byte(hi: u8, lo: u8) -> Result<u8, InvalidCharError> {
-    let hih = (hi as char).to_digit(16).ok_or(InvalidCharError { invalid: hi })?;
-    let loh = (lo as char).to_digit(16).ok_or(InvalidCharError { invalid: lo })?;
+fn hex_chars_to_byte(hi: u8, lo: u8) -> Result<u8, (u8, bool)> {
```
Comment (tcharding) [inline src/iter.rs:184]:
```suggestion
/// `hi` and `lo` are bytes representing hex characters.
///
/// Returns the valid byte or the invalid input byte and a bool indicating error for `hi` or `lo`.
fn hex_chars_to_byte(hi: u8, lo: u8) -> Result<u8, (u8, bool)> {
```

**Review by tcharding — APPROVED** [review]
Thanks for your contribution!

ACK 10571c434940951c481b39a77a8ff304e34bce52

My two suggestions are nits only, feel free to use or not.

**Review by tcharding — APPROVED** [review]
ACK 67556b3617ed7ff9b3e89f03e99a4e16a4f13db2

**Comment by tcharding** [comment]
Nice one, thanks man.

**Comment by leo-nash** [comment]
TIL `cargo clippy --tests`

**Comment by tcharding** [comment]
`cargo clippy --all-targets` would get you there as well.

**Review by tcharding — APPROVED** [review]
ACK d7b4be55e2fd17eeb7be60ebfc6ad52ada88067f

**Comment by apoelstra** [comment]
The `ExactSizeIterator` bound is actually pretty onerous. It means that users can no longer chain iterators, for example, since [`Chain` disables the trait for annoying reasons](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html).

I'm ok with this if Kix is (this isn't too hard to work around; typically you are only chaining iterators within encapsulated code where you can write your own non-crippled `Chain`). But want to draw attention to it.

**Review by apoelstra — APPROVED** [review]
ACK d7b4be55e2fd17eeb7be60ebfc6ad52ada88067f successfully ran local tests

**Comment by Kixunil** [comment]
@apoelstra thanks for pointing that out! I've just realized the iterator is actually publicly constructible, I haven't noticed the public constructor before, so I was thinking it's not a big deal to change it.

We could roll our own poor-man's specialization as `serde_json` does. Do we want to?

**Comment by apoelstra** [comment]
> We could roll our own poor-man's specialization as `serde_json` does. Do we want to?

You mean, having separate method names for exact-sized vs non-exact-sized iterators? Or something nicer than that?

Personally I think we shoud just accept this PR and accept that we have to do a major rev because of it (though we shouldn't do a release quickly until we've audited the rest of our errors at least).

**Comment by Kixunil** [comment]
> having separate method names for exact-sized vs non-exact-sized iterators? Or something nicer than that?

Possibly both, but I mainly had `&str` optimization in mind. I've also realized that my originally suggested approach to handle `char` doesn't work so we probably need this specialization anyway for it.

> Personally I think we shoud just accept this PR and accept that we have to do a major rev because of it

Yes, I think that blocking the PR over this is too much. I was mainly asking because we might want to revert to the original design instead.

**Comment by apoelstra** [comment]
> Possibly both, but I mainly had `&str` optimization in mind.

You mean, having separate methods for `&str` and for `&[u8]`?

>Yes, I think that blocking the PR over this is too much.

Ok, cool. I'll go ahead and merge it.

**Comment by Kixunil** [comment]
> You mean, having separate methods for `&str` and for `&[u8]`?

Yes.

**Review by Kixunil — APPROVED** [review]
ACK d7b4be55e2fd17eeb7be60ebfc6ad52ada88067f

**Pull request merged**